src/solver/take/forward_take.cpp

```diff
@@ -0,0 +1,124 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/take/invoke_params.hpp>
+#include <miopen/take/solvers.hpp>
+#include <miopen/take.hpp>
+#include <miopen/target_properties.hpp>
+
+// TODO: as my function is in 2024H2, I don't have test examples now
+// I have tuned this param with random tests I generated
+#define LOCAL_SIZE 1024
+
+namespace miopen {
+
+namespace solver {
+
+namespace take {
+
+bool TakeForward::IsApplicable(const ExecutionContext& context,
+                               const miopen::take::ProblemDescription& problem) const
+{
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsRightLength())
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    return true;
+}
+
+ConvSolution TakeForward::GetSolution(const ExecutionContext& context,
+                                      const miopen::take::ProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto input_dtype  = miopen::GetDataType(problem.GetXDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetYDesc().GetType());
+    auto ydims        = problem.GetYDesc().GetLengths();
+    auto output_numel =
+        std::accumulate(ydims.begin(), ydims.end(), 1ULL, std::multiplies<size_t>());
+
+    {
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(output_numel, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel = KernelInfo{};
+
+        kernel.kernel_file = "MIOpenTake.cpp";
+        kernel.kernel_name = "TakeFwdContiguous";
+
+        const auto build_params = KernelBuildParameters{
+            {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+            {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
```

Comment:
Since the types of input and output are the same, You just need to build it with one type.

Reply:
I saw that on [this PR](https://github.com/ROCm/MIOpen/pull/2583/files#diff-b890a7c8943e22e0e1e3ec36ed8892eb1f44706b32e460ce2bae566187b7eb9f) implementing argmax by @seungmanhan the types of input and output are also same but he used both INPUT_TYPE and OUTPUT_TYPE. Am I correct? Moreover, if I use one type like what you said, I should also change code in `driver/dm_take.cpp` from `TakeDriver<float16, float>()` to `TakeDriver<float16, float16>() and `TakeDriver<bfloat16, float>()` to `TakeDriver<float, float>()`, am I right?

Reply:
I personally think it's not a big deal. But if AMD had pointed out to fix that, I suggest you to follow that.
@et16kr Did AMD do that? Otherwise i think we can just pass this issue.

Reply:
Yes, in modnn, if it is the same type, usually only one DTYPE is entered.
I found out later that in miopen, there are cases where the input type and output type are separate even for the same type.
There seems to be no problem.